src/txParsers.ts

```diff
@@ -0,0 +1,282 @@
+        unparsedPoolParams,
+        ParseErrorReason.INVALID_POOL_PARAMS,
+        parsePoolKeyHash,
+        createParser(parseBufferOfLength, VRF_KEY_HASH_LENGTH, ParseErrorReason.INVALID_VRF_KEY_HASH),
+        parseCoin,
+        parseCoin,
+        parseUnitInterval,
+        parseRewardAccount,
+        createParser(parseArray, createParser(parseBufferOfLength, ADDRESS_KEY_HASH_LENGTH, ParseErrorReason.INVALID_POOL_OWNER), ParseErrorReason.INVALID_POOL_OWNERS),
+        createParser(parseArray, parseRelay, ParseErrorReason.INVALID_RELAYS),
+        createParser(parseNullable, parsePoolMetadata),
+    )
+
+    return {operator, vrfKeyHash, pledge, cost, margin, rewardAccount, poolOwners, relays, poolMetadata}
+}
+
+const parseCertificateType = (unparsedCertificateType: unknown): CertificateType => {
+    validate(isNumber(unparsedCertificateType), ParseErrorReason.INVALID_CERTIFICATE_TYPE)
+    validate(unparsedCertificateType in CertificateType, ParseErrorReason.INVALID_CERTIFICATE_TYPE)
+    return unparsedCertificateType
+}
+
+const parseStakeRegistrationCertificate = (data: unknown[]): WithoutType<StakeRegistrationCertificate> => ({
+    stakeCredential: parseStakeCredential(data[0]),
+})
+
+const parseStakeDeregistrationCertificate = (data: unknown[]): WithoutType<StakeDeregistrationCertificate> => ({
+    stakeCredential: parseStakeCredential(data[0]),
+})
+
+const parseStakeDelegationCertificate = (data: unknown[]): WithoutType<StakeDelegationCertificate> => ({
+    stakeCredential: parseStakeCredential(data[0]),
+    poolKeyHash: parsePoolKeyHash(data[1]),
+})
+
+const parsePoolRegistrationCertificate = (data: unknown[]): WithoutType<PoolRegistrationCertificate> => ({
+    poolParams: parsePoolParams(data),
+})
+
+const parsePoolRetirementCertificate = (data: unknown[]): WithoutType<PoolRetirementCertificate> => ({
+    poolKeyHash: parsePoolKeyHash(data[0]),
+    epoch: parseEpoch(data[1]),
+})
+
+const parseGenesisKeyDelegation = (data: unknown[]): WithoutType<GenesisKeyDelegation> => ({
+    restOfData: data,
+})
+
+const parseMoveInstantaneousRewardsCertificate = (data: unknown[]): WithoutType<MoveInstantaneousRewardsCertificate> => ({
+    restOfData: data,
+})
+
+const parseCertificate = createParser(
+    parseBasedOnType,
+    ParseErrorReason.INVALID_CERTIFICATE,
+    parseCertificateType,
+    parseStakeRegistrationCertificate,
+    parseStakeDeregistrationCertificate,
+    parseStakeDelegationCertificate,
+    parsePoolRegistrationCertificate,
+    parsePoolRetirementCertificate,
+    parseGenesisKeyDelegation,
+    parseMoveInstantaneousRewardsCertificate,
+)
+
+export const parseInputs = createParser(parseArray, parseTxInput, ParseErrorReason.INVALID_TX_INPUTS)
+export const parseOutputs = createParser(parseArray, parseTxOutput, ParseErrorReason.INVALID_TX_OUTPUTS)
+const parseFee = createParser(parseUint, ParseErrorReason.INVALID_TX_FEE)
```

Comment:
Why are these not exported?

Reply:
I exported the other ones so that I was able to write "unit" tests just for parsing various inputs, outputs, ... separately, e.g. not in a transaction body so I needed to export them. I didn't consider it necessary to export ones as `parseFee` as they are just aliases for simple parsers such as `parseUint`.

Reply:
I guess I'd export them all... even if not really needed at the moment. If I see, say, parseMetadataHash and want to write analogously parseTtl, it would fail annoyingly.